bee-message/src/unlock/mod.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+mod reference;
+
+pub use reference::ReferenceUnlock;
+
+use crate::{constants::UNLOCK_BLOCK_COUNT_RANGE, error::ValidationError, signature::SignatureUnlock};
+
+use bee_packable::{
+    error::{PackPrefixError, UnpackPrefixError},
+    Packable, UnknownTagError, VecPrefix,
+};
+
+use core::{convert::Infallible, ops::Deref};
+use std::collections::HashSet;
+
+/// Defines the mechanism by which a transaction input is authorized to be consumed.
+#[non_exhaustive]
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Packable)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(tag = "type", content = "data")
+)]
+#[packable(tag_type = u8)]
+pub enum UnlockBlock {
+    /// A signature unlock block.
+    #[packable(tag = 0)]
+    Signature(SignatureUnlock),
+    /// A reference unlock block.
+    #[packable(tag = 1)]
+    Reference(ReferenceUnlock),
+}
+
+impl UnlockBlock {
+    /// Returns the unlock kind of an `UnlockBlock`.
+    pub fn kind(&self) -> u8 {
+        match self {
+            Self::Signature(_) => SignatureUnlock::KIND,
+            Self::Reference(_) => ReferenceUnlock::KIND,
+        }
+    }
+}
+
+impl From<SignatureUnlock> for UnlockBlock {
+    fn from(signature: SignatureUnlock) -> Self {
+        Self::Signature(signature)
+    }
+}
+
+impl From<ReferenceUnlock> for UnlockBlock {
+    fn from(reference: ReferenceUnlock) -> Self {
+        Self::Reference(reference)
+    }
+}
+
+/// A collection of unlock blocks.
+#[derive(Clone, Debug, Eq, PartialEq, Packable)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[packable(pack_error = PackPrefixError<Infallible, u16>)]
+#[packable(unpack_error = UnpackPrefixError<UnknownTagError<u8>, u16>)]
+pub struct UnlockBlocks {
+    #[packable(wrapper = VecPrefix<UnlockBlock, u16>)]
+    inner: Vec<UnlockBlock>,
+}
+
+impl UnlockBlocks {
+    /// Creates a new `UnlockBlocks`.
+    pub fn new(unlock_blocks: Vec<UnlockBlock>) -> Result<Self, ValidationError> {
+        if !UNLOCK_BLOCK_COUNT_RANGE.contains(&unlock_blocks.len()) {
+            return Err(ValidationError::InvalidUnlockBlockCount(unlock_blocks.len()));
+        }
+
+        let mut seen_signatures = HashSet::new();
+
+        for (index, unlock_block) in unlock_blocks.iter().enumerate() {
+            match unlock_block {
+                UnlockBlock::Reference(r) => {
+                    if index == 0
+                        || r.index() >= index as u16
```

Comment:
is there any reason to believe this truncation could produce undesired effects?

Reply:
`index` will always be a valid `u16` here, since the validation checks that the length of the `UnlockBlock` vector is <= 127 beforehand, so there shouldn't be any unexpected truncations here.